c++/src/kj/async-io.h

```diff
@@ -136,6 +136,60 @@ class AsyncOutputStream: private AsyncObject {
   // multiple times without canceling the previous promises.
 };
 
+class AsyncBufferedInputStream {
+  // Does not inherit from InputStream so this can act as a mixin to allow implementing
+  // AsyncBufferedIoStreamWrapper
+
+public:
+  virtual ~AsyncBufferedInputStream() noexcept(false);
+
+  ArrayPtr<const byte> getReadBuffer();
+  // TODO(now): No skip for AsyncInputStream -- I guess we could implement an async form?
+  // Get a direct pointer into the read buffer, which contains the next bytes in the input.  If the
+  // caller consumes any bytes, it should then call skip() to indicate this.  This always returns a
+  // non-empty buffer or throws an exception.  Implemented in terms of tryGetReadBuffer().
+
+  virtual ArrayPtr<const byte> tryGetReadBuffer() = 0;
+  // Like getReadBuffer() but may return an empty buffer on EOF.
+};
+
+class AsyncBufferedInputStreamWrapper: public AsyncBufferedInputStream, public AsyncInputStream {
+  // Implements AsyncBufferedInputStream in terms of AsyncInputStream
+  //
+  // Note that the underlying stream's position is unpredictable once the wrapper is destroyed,
+  // unless the entire stream was consumed.  To read a predictable number of bytes in a buffered
+  // way without going over, you'd need this wrapper to wrap some other wrapper which itself
+  // implements an artificial EOF at the desired point.  Such a stream should be trivial to write
+  // but is not provided by the library at this time.
+
+public:
+  explicit AsyncBufferedInputStreamWrapper(AsyncInputStream& inner, ArrayPtr<byte> buffer = nullptr);
+  // Creates a buffered stream wrapping the given non-buffered stream.  No guarantee is made about
+  // the position of the inner stream after a buffered wrapper has been created unless the entire
+  // input is read.
+  //
+  // If the second parameter is non-null, the stream uses the given buffer instead of allocating
+  // its own.  This may improve performance if the buffer can be reused.
+
+  AsyncBufferedInputStreamWrapper(AsyncBufferedInputStreamWrapper&& other);
```

Comment:
Why is there a move constructor? That seems unexpected for this kind of class.

Reply:
I wanted to use this class with `kj::attachVal` to tie together ownership of the inner stream / buffer where applicable. I guess you can `kj::heap` the wrapper if that seems weird, or implement the pattern where there are varations of the constructor that own the stream and/or buffer and add ownStream/ownBuffer class members.

Reply:
Yeah I would have a constructor that allows passing in an owned buffer.

It feels generally weird for a type that has async methods to be movable, since it's possibly ambiguous whether you're allowed to move it while a promise is outstanding. KJ style guide says that resource types shouldn't be movable for reasons like this.